### DISCUSSION TITLE
mitmproxy works in transparent mode, regardless of how it was started

### DISCUSSION THREAD

**Discussion by omicronns** [mitmproxy/mitmproxy#6506] [General]
Hi,

I was trying to run `mitmproxy` in transparent mode on other machine than the routing is done (Raspberry Pi running openwrt and my laptop), but I was getting errors: `Transparent mode failure: OSError(92, 'Protocol not available')`. Some time into debugging i started `mitmweb` in regular mode (just by running `mitmweb`) by mistake, and it started working correctly. I inspected traffic coming to `mitmweb` proxy with wireshark and there are no `HTTP CONNECT` packets, just straight TLS data. When I disabled port forwarding rules on raspberry and just used regular proxy configuration to connect it to my laptop running `mitmweb` it also worked but `HTTP CONNECT` packets are now present.

Is this expected behavior?

**Comment by mhils** [comment]
> Transparent mode failure: OSError(92, 'Protocol not available')

That looks like an error with your transparent proxy configuration. You should find some hits in our issue tracker for that error.

> Is this expected behavior?

The rest you described is expected behavior. In regular mode, mitmproxy will route based on the HTTP Host header if there is no CONNECT request. This means things will often "just work" in regular mode, but you need something else for non-HTTP/missing host headers/... . Does that make sense? :)

**Reply by omicronns** [reply]
> Does that make sense?

So when mitmproxy is spawned in regular mode and HTTP proxy protocol is used, information about routing is source address of ip packet and content of CONNECT packet. When firewall is used to redirect packets into mitmproxy instead of proxy configuration on client, then mitmproxy falls back to to using source ip packet address/Host header for routing.

On the other hand when mitmproxy is spawned in transparent mode both source and destination address of ip packets are used for routing?
